src/data/matrix/kronecker_product_bialgebra.lean

```diff
@@ -0,0 +1,198 @@
+  (kronecker_biprod α R S β A B) := by { ext ⟨i, i'⟩ ⟨j, j'⟩, refl }
+
+lemma kronecker_biprod_reindex_right (eₙ : n ≃ n') (eₚ : p ≃ p') (A : matrix l m R)
+  (B : matrix n p S) [is_scalar_tower α R β] [is_scalar_tower α S β] :
+  kronecker_biprod α R S β A (reindex_linear_equiv α S eₙ eₚ B) =
+  reindex_linear_equiv α _ ((equiv.refl _).prod_congr eₙ) ((equiv.refl _).prod_congr eₚ)
+  (kronecker_biprod α R S β A B) := by { ext ⟨i, i'⟩ ⟨j, j'⟩, refl }
+
+lemma kronecker_biprod_one_one [decidable_eq m] [decidable_eq n] [is_scalar_tower α R β]
+  [is_scalar_tower α S β] : kronecker_biprod α R S β (1 : matrix m m R)
+    (1 : matrix n n S) = (1 : matrix (m × n) (m × n) β) :=
+    by { ext ⟨i, i'⟩ ⟨j, j'⟩, simp [kronecker_biprod, one_apply, algebra_map_eq_smul_one, ite_smul,
+      ite_and] }
+
+theorem kronecker_biprod_mul (A : matrix l m R) (B : matrix m n R) (A' : matrix l' m' S)
+  (B' : matrix m' n' S) [is_scalar_tower α R β] [is_scalar_tower α S β] :
+  kronecker_biprod α R S β (A ⬝ B) (A' ⬝ B') =
+    (kronecker_biprod α R S β A A') ⬝ (kronecker_biprod α R S β B B') :=
+begin
+  ext ⟨i, i'⟩ ⟨j, j'⟩,
+  simp only [mul_apply, kronecker_biprod_apply_apply, algebra_map_eq_smul_one, mul_one,
+    algebra.mul_smul_comm,
+    linear_map.coe_mk, algebra.smul_mul_assoc, ← finset.univ_product_univ, finset.sum_product,
+    kronecker_map, algebra.biprod_apply],
+  simp_rw [finset.sum_smul, finset.smul_sum, ← smul_eq_mul],
+  repeat {apply finset.sum_congr, refl, intros _ _,},
+  rw [is_scalar_tower.smul_assoc, id.smul_eq_mul (A' i' x_1) (B' x_1 j')],
+  simp only [id.smul_eq_mul, ← algebra_map_eq_smul_one, smul_def, ring_hom.map_mul],
+  ring_nf,
+end
+
+theorem kronecker_biprod_assoc {T : Type*} [comm_semiring T] [algebra α T] [algebra T β]
+ [is_scalar_tower α R β] [is_scalar_tower α S β] [is_scalar_tower α T β] [is_scalar_tower α β β]
+ (A : matrix m m' R) (B : matrix n n' S) (C : matrix p p' T) :
+  @matrix.linear_equiv_index_assoc m n p _ _ _ m' n' p' _ _ _ β α _ _ _
+  (kronecker_biprod α β T β (kronecker_biprod α R S β A B) C) =
+    kronecker_biprod α R β β A (kronecker_biprod α S T β B C)
+    :=
+begin
+  simp only [matrix.linear_equiv_index_assoc, kronecker_biprod_apply_apply, linear_map.coe_mk,
+    id.map_eq_self,
+    reindex_apply, linear_equiv.coe_mk],
+  ext i j,
+  simp only [equiv.prod_assoc_symm_apply, reindex_linear_equiv_apply, minor_apply, reindex_apply,
+    mul_assoc, kronecker_map, algebra.biprod_apply, id.map_eq_self],
+end
+
+end matrix_bialgebra
+
+namespace kronecker_product
+
+open algebra matrix matrix_bialgebra
+open_locale matrix
+
+variables (R : Type*) [comm_semiring R]
+variables {l m n p l' m' n' p' : Type*}
+variables [fintype l] [fintype m] [fintype n] [fintype p]
+variables [fintype l'] [fintype m'] [fintype n'] [fintype p']
+
+def kronecker_prod (A : matrix l m R) (B : matrix n p R) :
+  matrix (l × n) (m × p) R := kronecker_biprod R R R R A B
```

Comment:
If you define this as 
```suggestion
  matrix (l × n) (m × p) R := kronecker_map (*) A B
```
Then it generalizes to `has_mul R`

Reply:
Ok, thanks; I will generalize it.

Reply:
Well, actually there is a small problem, I think. The way all these proofs are set-up is to adapt kronecker_biprod, and there I need all guys to be comm_semirings.

Reply:
You can still take stronger assumptions on the lemmas about this definition that need them - my point was just that the definition didn't need those assumptions.

Reply:
I see; yet, I personally prefer the older version since it seems more natural (to me) from a mathematical point of view, as a "special case of a biproduct". But if you insist I can change the definition.

Reply:
Let's leave it as you have it for now, I'll take another look at this in a few days once FMM is over. As a compromise, you can at least relax this to non-commutative rings by replacing the first R with `ℕ`.

Reply:
Well, the problem is that for `kronecker_biprod` I needed R to be an \alpha-algebra, and this forces \alpha to be commutative. No?

Reply:
Sorry, I meant pick alpha as `nat` - I might have had the argument order wrong.

Reply:
I strongly suspect that `\beta` does not need to be an R and S algebra, but I admit that this is now more complex that the one character change I was suggesting

Reply:
Ok, I see. So if I understand correctly we leave this PR on hold until the FMM (which I googled for: Franciscan Missionaries of Mary? Federation of Malaysian Manufacturers? Florida Maritime Museum? For all of these, to see them over seems a bit sad...) is over, right? I would happily move the `TODO: move` part elsewhere, in the meanwhile, and understand why `bors` insists that this `PR` depends on another one.

Reply:
[FMM](https://cicm-conference.org/2021/cicm.php?event=fmm)

Reply:
Oh, I wasn't aware of the conference, thanks!